diff --git a/dist/chunk/4XGMYOCT.js b/dist/chunk/4XGMYOCT.js
index 91b8b266c385a1795286b0bc49253c67389d7dff..bc8a7b2b7f9b73e351adc5509a9508c60d603059 100644
--- a/dist/chunk/4XGMYOCT.js
+++ b/dist/chunk/4XGMYOCT.js
@@ -10,7 +10,7 @@ import { mergeRefs, composeEventHandlers, mergeDefaultProps, getFocusableTreeWal
 import { createContext, useContext, createSignal, splitProps, createEffect, on, onCleanup, Show, onMount, createUniqueId, createMemo, mergeProps as mergeProps$1 } from 'solid-js';
 import createPresence from 'solid-presence';
 import { combineStyle } from '@solid-primitives/props';
-import { createResizeObserver } from '@solid-primitives/resize-observer';
+import { makeResizeObserver } from '@solid-primitives/resize-observer';
 
 // src/tabs/index.tsx
 var tabs_exports = {};
@@ -136,19 +136,14 @@ function TabsIndicator(props) {
       computeStyle();
     });
   });
-  createEffect(on([context.selectedTab, context.orientation, direction], () => {
+  createEffect(on([context.selectedTab, context.orientation, direction, () => context.listState().collection()], () => {
     computeStyle();
   }, {
     defer: true
   }));
   const [resizing, setResizing] = createSignal(false);
   let timeout = null;
-  let prevTarget = null;
-  createResizeObserver(context.selectedTab, (_, t) => {
-    if (prevTarget !== t) {
-      prevTarget = t;
-      return;
-    }
+  const { observe, unobserve } = makeResizeObserver(() => {
     setResizing(true);
     if (timeout)
       clearTimeout(timeout);
@@ -157,6 +152,14 @@ function TabsIndicator(props) {
       setResizing(false);
     }, 1);
     computeStyle();
+  }, { box: "border-box" });
+  createEffect(() => {
+    for (const item of context.listState().collection()) {
+      const tab = item.rawValue.ref();
+      if (!tab) continue;
+      observe(tab);
+      onCleanup(() => unobserve(tab));
+    }
   });
   return createComponent(Polymorphic, mergeProps({
     as: "div",
diff --git a/dist/chunk/ZABX5HRE.jsx b/dist/chunk/ZABX5HRE.jsx
index 0d295ad548605fd85c324b1ff74aae123341b0a3..f590daa378880836644c7cf030935bbf32a45631 100644
--- a/dist/chunk/ZABX5HRE.jsx
+++ b/dist/chunk/ZABX5HRE.jsx
@@ -132,7 +132,7 @@ import {
   splitProps as splitProps2
 } from "solid-js";
 import { combineStyle } from "@solid-primitives/props";
-import { createResizeObserver } from "@solid-primitives/resize-observer";
+import { makeResizeObserver } from "@solid-primitives/resize-observer";
 function TabsIndicator(props) {
   const context = useTabsContext();
   const [local, others] = splitProps2(props, ["style"]);
@@ -167,7 +167,7 @@ function TabsIndicator(props) {
   });
   createEffect2(
     on2(
-      [context.selectedTab, context.orientation, direction],
+      [context.selectedTab, context.orientation, direction, () => context.listState().collection()],
       () => {
         computeStyle();
       },
@@ -176,12 +176,7 @@ function TabsIndicator(props) {
   );
   const [resizing, setResizing] = createSignal2(false);
   let timeout = null;
-  let prevTarget = null;
-  createResizeObserver(context.selectedTab, (_, t) => {
-    if (prevTarget !== t) {
-      prevTarget = t;
-      return;
-    }
+  const { observe, unobserve } = makeResizeObserver(() => {
     setResizing(true);
     if (timeout)
       clearTimeout(timeout);
@@ -190,6 +185,14 @@ function TabsIndicator(props) {
       setResizing(false);
     }, 1);
     computeStyle();
+  }, { box: "border-box" });
+  createEffect2(() => {
+    for (const item of context.listState().collection()) {
+      const tab = item.rawValue.ref();
+      if (!tab) continue;
+      observe(tab);
+      onCleanup(() => unobserve(tab));
+    }
   });
   return <Polymorphic
     as="div"
diff --git a/src/tabs/tabs-indicator.tsx b/src/tabs/tabs-indicator.tsx
index f7fc714539efd59b01dd6e2da5ec7f162b53c84f..9f209c6e863678815cccddf10abc14563d9f4541 100644
--- a/src/tabs/tabs-indicator.tsx
+++ b/src/tabs/tabs-indicator.tsx
@@ -12,12 +12,13 @@ import {
 	createSignal,
 	on,
 	onMount,
+	onCleanup,
 	splitProps,
 } from "solid-js";
 
 import type { Orientation } from "@kobalte/utils";
 import { combineStyle } from "@solid-primitives/props";
-import { createResizeObserver } from "@solid-primitives/resize-observer";
+import { makeResizeObserver } from "@solid-primitives/resize-observer";
 import { useLocale } from "../i18n";
 import {
 	type ElementOf,
@@ -107,7 +108,7 @@ export function TabsIndicator<T extends ValidComponent = "div">(
 	// Compute style normally for subsequent runs.
 	createEffect(
 		on(
-			[context.selectedTab, context.orientation, direction],
+			[context.selectedTab, context.orientation, direction, () => context.listState().collection()],
 			() => {
 				computeStyle();
 			},
@@ -118,13 +119,7 @@ export function TabsIndicator<T extends ValidComponent = "div">(
 	const [resizing, setResizing] = createSignal(false);
 
 	let timeout: NodeJS.Timeout | null = null;
-	let prevTarget: any = null;
-	createResizeObserver(context.selectedTab, (_, t) => {
-		if (prevTarget !== t) {
-			prevTarget = t;
-			return;
-		}
-
+	const { observe, unobserve } = makeResizeObserver(() => {
 		setResizing(true);
 
 		if (timeout) clearTimeout(timeout);
@@ -136,6 +131,15 @@ export function TabsIndicator<T extends ValidComponent = "div">(
 		}, 1);
 
 		computeStyle();
+	}, { box: "border-box" });
+
+	createEffect(() => {
+		for (const item of context.listState().collection()) {
+			const tab = item.rawValue.ref();
+			if (!tab) continue;
+			observe(tab);
+			onCleanup(() => unobserve(tab));
+		}
 	});
 
 	return (
